### DISCUSSION TITLE
Laravel Filament Speed issue.

### DISCUSSION THREAD

**Discussion by Shaunshapeshift** [filamentphp/filament#13426] [Help]
### Package

Other

### Package Version

v3

### How can we help you?

I am currently working on a project for a Client, utilizing Filament in combination with Laravel to build a robust survey template builder that supports multiple languages. We have encountered some performance issues and would greatly appreciate your assistance in optimizing the processing time.

We are currently facing significant performance issues, particularly with loading times and responsiveness of the interface.

We are seeking your expertise to help us optimize our implementation. Specifically, we would like to understand: best practices, performance optimization and caching strategies. We appreciate your assistance in resolving these issues and look forward to your guidance.

Here is a code implementation:
We use JSON columns to store language translations for the names of regions, markets, brands, and sub-brands. Below is an example of the implementation within the DistributorResource class.
namespace App\Filament\Resources;

use App\Models\Brand;
use App\Models\Category;
use App\Models\Distributor;
use App\Models\Market;
use App\Models\Region;
use App\Models\SubBrand;
use Filament\Forms;
use Filament\Forms\Form;
use Filament\Forms\Get;
use Filament\Forms\Set;
use Filament\Resources\Resource;
use Filament\Tables;
use Filament\Tables\Table;

class DistributorResource extends Resource
{
    protected static ?string $model = Distributor::class;
    protected static ?string $navigationIcon = 'heroicon-o-rectangle-stack';
    protected static ?string $recordTitleAttribute = 'name';
    protected static ?string $navigationLabel = 'Distributor List';
    protected static ?int $navigationSort = 3;

    public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Forms\Components\TextInput::make('name')->required()->maxLength(255),
                Forms\Components\Select::make('regions')->required()->multiple()->relationship('regions', 'name')
                    ->options(fn () => Region::all()->mapWithKeys(
                        fn ($region) => [$region->id => $region->name['en']]
                    ))
                    ->getSearchResultsUsing(fn (string $search) => Region::where('name', 'like', "%{$search}%")->get()->mapWithKeys(
                        fn ($region) => [$region->id => $region->name['en']]
                    ))
                    ->afterStateUpdated(fn (Set $set) => $set('markets', []))
                    ->placeholder('Select regions')->live(),
                Forms\Components\Select::make('markets')->required()->multiple()->relationship('markets', 'name')->preload()
                    ->options(fn (Get $get) => Market::whereIn('region_id', $get('regions') ?? [])->get()->mapWithKeys(
                        fn ($market) => [$market->id => $market->name['en']]
                    ))
                    ->getSearchResultsUsing(fn (string $search, Get $get) => Market::where('name', 'like', "%$search%")->whereIn('region_id', $get('regions') ?? [])->get()->mapWithKeys(
                        fn ($market) => [$market->id => $market->name['en']]
                    ))
                    ->live()->placeholder('Select markets'),
                Forms\Components\TextInput::make('company_registration_number')->label('Company Registration Number')->maxLength(255),
                Forms\Components\TextInput::make('contact_person')->label('Contact Person')->required()->maxLength(255),
                Forms\Components\TextInput::make('contact_email')->label('Contact Email')->required()->maxLength(255),
                Forms\Components\Select::make('categories')->label('Categories')->relationship('categories', 'name')->multiple()
                    ->options(fn () => Category::all()->mapWithKeys(
                        fn ($category) => [$category->id => $category->name['en']]
                    ))
                    ->getSearchResultsUsing(fn (string $search) => Category::where('name', 'like', "%{$search}%")->get()->mapWithKeys(
                        fn ($category) => [$category->id => $category->name['en']]
                    ))
                    ->live()->afterStateUpdated(fn (Set $set) => $set('brands', []))->afterStateUpdated(fn (Set $set) => $set('subBrands', []))->placeholder('Select categories')->searchable(false),
                Forms\Components\Select::make('brands')->label('Marketed Coty Brands')->relationship('brands', 'name')->multiple()
                    ->options(fn (Get $get) => Brand::whereHas(
                        'categories',
                        fn ($q) => $q->whereIn('id', $get('categories') ?? []))->get()->mapWithKeys(
                            fn ($brand) => [$brand->id => $brand->name['en']]
                        )
                    )
                    ->getSearchResultsUsing(fn (string $search, Get $get) => Brand::where('name', 'like', "%{$search}%")->whereHas(
                        'categories',
                        fn ($q) => $q->whereIn('id', $get('categories') ?? []))->get()->mapWithKeys(
                            fn ($brand) => [$brand->id => $brand->name['en']]
                        )
                    )
                    ->preload()->live()->afterStateUpdated(fn (Set $set) => $set('subBrands', []))->searchable(false)->placeholder('Select brands'),
                Forms\Components\Select::make('subBrands')->label('Sub Brands')->relationship('subBrands', 'name')->multiple()
                    ->options(fn (Get $get) => SubBrand::whereIn('brand_id', $get('brands') ?? [1, 2, 3])->get()->mapWithKeys(
                        fn ($brand) => [$brand->id => $brand->name['en']]
                    ))
                    ->getSearchResultsUsing(fn (string $search, Get $get) => SubBrand::where('name', 'like', "%{$search}%")->whereIn('brand_id', $get('brands') ?? [1, 2, 3])->get()->mapWithKeys(
                        fn ($brand) => [$brand->id => $brand->name['en']]
                    ))
                    ->preload()->live()->searchable(false)->placeholder('Select sub brands'),
            ])->columns(3);
    }

    public static function table(Table $table): Table
    {
        return $table
            ->columns([
                Tables\Columns\TextColumn::make('name')->label('Distributor Name'),
                Tables\Columns\TextColumn::make('regions')->label('Regions')->formatStateUsing(function ($record) {
                    $locale = app()->getLocale();

                    return $record->regions->pluck("name.$locale")->implode(', ');
                }),
                Tables\Columns\TextColumn::make('markets')->label('Markets')->formatStateUsing(function ($record) {
                    $locale = app()->getLocale();

                    return $record->markets->pluck("name.$locale")->implode(', ');
                }),
                Tables\Columns\TextColumn::make('contact_person')->label('Contact Person'),
                Tables\Columns\TextColumn::make('contact_email')->label('Contact Email'),
            ]);
    }
}

**Comment by PovilasKorop** [comment]
@Shaunshapeshift hi, interesting case, I would love to debug it and give any advice for others in my audience on YouTube channel Filament Daily.

To do that, I would need to reproduce the situation and play around with it.

Any chance you would give me the access to the repository? I wouldn't show the code anywhere in public.

Alternatively, I can try to reproduce the situation myself but then it would take me a few days until I find enough free time for it.

Also, can you show which exactly part is slow or which action? Any benchmarks or screencasts?
